src/storage/src/render/sources.rs

```diff
@@ -201,15 +201,30 @@ where
     } = description.desc;
 
     let (decoded_stream, decode_health) = match encoding {
-        None => (
-            ok_source.map(|r| DecodeResult {
-                key: None,
-                value: Some(Ok(r.value)),
-                metadata: Row::default(),
-                from_time: r.from_time,
-            }),
-            empty(scope),
-        ),
+        None => {
+            (
+                ok_source.map(|r| {
+                    let mut key = r.key.iter();
+                    let first = key.next();
+                    let more = key.next().is_some();
+                    DecodeResult {
+                        // The rest of the rendering code (particularly upsert) relies on the fact
+                        // that `Null`'s provided by the source are mapped to `None`'s.
+                        //
+                        // TODO(guswynn|petrosagg): these constraints should be lifted, or rendering
+                        // should use native `Datum::Null` representations and not this option-wrapping.
+                        key: match (first, more) {
```

Comment:
This is making a rendering decision based on the schema of the row. This is never ok and always indicating that some decision must be taken in the planner and encoded in the source description. In this particular case the planner needs to decide which way this mapping should go and encode this fact in a flag or something.